### PR TITLE
Further CUDA support for IF-1

### LINKED ISSUE #374: Configuring device in Inverse Folding model
Hello,

This is related to https://github.com/facebookresearch/esm/pull/317 ; none of the main functionalities of esm_if1_gvp4_t16_142M_UR50 seem to run on GPU. This applies to model.sample ; esm.inverse_folding.util.score_sequence and esm.inverse_folding.util.get_encoder_output.

```
from copy import deepcopy
import numpy as np
import torch
import torch.nn.functional as F
import torch_geometric
import torch_sparse
from torch_geometric.nn import MessagePassing
import esm
from esm.inverse_folding.util import CoordBatchConverter

cuda_devices = 10
device = torch.device(f"cuda:{cuda_devices}" if (torch.cuda.is_available()) else "cpu")
model, alphabet = esm.pretrained.esm_if1_gvp4_t16_142M_UR50()
model = model.eval()
model.to(device)

pdb_path =  "something.pdb"
structure = esm.inverse_folding.util.load_structure(pdb_path)
coords, native_seq = esm.inverse_folding.util.extract_coords_from_structure(structure)


sampled_seq = model.sample(coords, temperature=1.)
ll_fullseq, ll_withcoord = esm.inverse_folding.util.score_sequence(model, alphabet, coords, native_seq)
rep = esm.inverse_folding.util.get_encoder_output(model, alphabet, coords)
# all 3 trigger "RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:10 and cpu! (when checking argument for argument index in method wrapper__index_select)"

model.to(torch.device("cpu"))
sampled_seq = model.sample(coords, temperature=1.)
ll_fullseq, ll_withcoord = esm.inverse_folding.util.score_sequence(model, alphabet, coords, native_seq)
rep = esm.inverse_folding.util.get_encoder_output(model, alphabet, coords)
# no problem
```

This is a bit unfortunate and can be resolved rather easily ... Do you plan to fix this please ? Or would you like a PR ?
If not, is there a specific reason not to run these methods on GPU ? We could do an allclose test and verify that results are consistent.

Thanks !

### PR DESCRIPTION
**Pull request by naailkhan28**
Resolves #374 - now `model.sample` takes an optional `device` argument. `util.get_encoder_output` and `multichain_util.sample_sequence_in_complex` will use the same device that the model is on already here.

### CODE REVIEW

**Review by tomsercu — APPROVED** [review]
Lgtm!

Would you be able to update the entrypoints `sample_sequences.py` and `score_log_likelihoods.py` to use the gpu when available? I prefer to have something like a `--nogpu` flag such that the default is the "reasonable" thing. Example: https://github.com/facebookresearch/esm/blob/main/scripts/extract.py#L59

**Comment by naailkhan28** [comment]
Updated `sample_sequences.py` and `score_log_likelihoods.py`, should be all set!

File: examples/inverse_folding/score_log_likelihoods.py [hunk 1]
```diff
@@ -110,6 +116,8 @@ def main():
 
     model, alphabet = esm.pretrained.esm_if1_gvp4_t16_142M_UR50()
     model = model.eval()
+    
+    parser.add_argument("--nogpu", action="store_true", help="Do not use GPU even if available")
```
Comment (brianhie) [inline examples/inverse_folding/score_log_likelihoods.py:120]: @naailkhan28 This needs to be moved up before the call to `args = parser.parse_args()`

**Review by brianhie — CHANGES_REQUESTED** [review]
When I run the command:
```
python examples/inverse_folding/score_log_likelihoods.py examples/inverse_folding/data/5YH2.pdb examples/inverse_folding/data/5YH2_mutated_seqs.fasta --chain C --outpath output/5YH2_mutated_seqs_scores.csv
```
with `--nogpu` disabled, I get the following error
```
Traceback (most recent call last):
  File "examples/inverse_folding/score_log_likelihoods.py", line 129, in <module>
    main()
  File "examples/inverse_folding/score_log_likelihoods.py", line 124, in main
    score_singlechain_backbone(model, alphabet, args)
  File "examples/inverse_folding/score_log_likelihoods.py", line 32, in score_singlechain_backbone
    ll, _ = esm.inverse_folding.util.score_sequence(
  File "/private/home/brianhie/ssf/third_party/esm_dev/esm/esm/inverse_folding/util.py", line 112, in score_sequence
    logits, _ = model.forward(coords, padding_mask, confidence, prev_output_tokens)
  File "/private/home/brianhie/ssf/third_party/esm_dev/esm/esm/inverse_folding/gvp_transformer.py", line 78, in forward
    encoder_out = self.encoder(coords, padding_mask, confidence,
  File "/private/home/brianhie/miniconda3/envs/paito/envs/ssf/envs/proteinseq_new/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/private/home/brianhie/ssf/third_party/esm_dev/esm/esm/inverse_folding/gvp_transformer_encoder.py", line 154, in forward
    x, encoder_embedding = self.forward_embedding(coords,
  File "/private/home/brianhie/ssf/third_party/esm_dev/esm/esm/inverse_folding/gvp_transformer_encoder.py", line 87, in forward_embedding
    components["tokens"] = self.embed_tokens(mask_tokens) * self.embed_scale
  File "/private/home/brianhie/miniconda3/envs/paito/envs/ssf/envs/proteinseq_new/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/private/home/brianhie/miniconda3/envs/paito/envs/ssf/envs/proteinseq_new/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 158, in forward
    return F.embedding(
  File "/private/home/brianhie/miniconda3/envs/paito/envs/ssf/envs/proteinseq_new/lib/python3.8/site-packages/torch/nn/functional.py", line 2183, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument index in method wrapper__index_select)
```

File: examples/inverse_folding/score_log_likelihoods.py (hunk 1 above)
Comment (naailkhan28) [reply examples/inverse_folding/score_log_likelihoods.py]: Thanks for spotting this! I've added a commit to fix this :)

**Review by brianhie — CHANGES_REQUESTED** [review]
@naailkhan28 I think scoring is still broken -- at least one of the tensors involved in the computation is still on CPU and it's giving me the error `RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`. It would be good to fix scoring as well, or at the very least just implement CUDA support for only sampling and remove the scoring support for now.

I'm testing with the command:
```
python examples/inverse_folding/score_log_likelihoods.py examples/inverse_folding/data/5YH2.pdb examples/inverse_folding/data/5YH2_mutated_seqs.fasta --chain C --outpath output/5YH2_mutated_seqs_scores.csv
```

**Comment by naailkhan28** [comment]
@brianhie I've added another commit that should fix this - it looks like it was an error in `inverse_folding.util.get_sequence_loss`

**Review by brianhie — APPROVED** [review]

**Comment by brianhie** [comment]
LGTM! Confirmed scoring between default and --nogpu, will merge. Thanks @naailkhan28 for the contribution!

**Pull request merged**